impl/src/features.rs

```diff
@@ -0,0 +1,482 @@
+}
+
+impl Features {
+  pub fn empty() -> Features {
+    Features {
+      features: Vec::new(),
+      targeted_features: vec![],
+    }
+  }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct TargetedFeatures {
+  pub platforms: Vec<String>,
+  pub features: Vec<String>,
+}
+
+// A function that runs `cargo-tree` to analyze per-platform features.
+// This step should not need to be separate from cargo-metadata, but cargo-metadata's
+// output is incomplete in this respect.
+//
+// See: https://github.com/rust-lang/cargo/issues/9863
+// and: https://github.com/illicitonion/cargo-metadata-pathologies/tree/main/platform-specific-features
+//
+// There is no plan for Cargo to expose platform data in `cargo-metadata`:
+// See https://github.com/rust-lang/cargo/pull/9982#issuecomment-965423533
+//
+// The Rust documentation says `cargo-tree` may stabilize its `--unit-graph` output, which
+// would make this function faster by avoiding mulitple invocations of `cargo-tree`.
+// https://rust-lang.github.io/rfcs/2957-cargo-features2.html#cargo-metadata
+//
+pub fn get_per_platform_features(
+  cargo_dir: &Path,
+  settings: &RazeSettings,
+  packages: &[Package],
+) -> Result<BTreeMap<PackageId, Features>> {
+  get_per_platform_features_with_command(cargo_dir, settings, packages, run_cargo_tree)
+}
+
+pub fn get_per_platform_features_with_command(
+  cargo_dir: &Path,
+  settings: &RazeSettings,
+  packages: &[Package],
+  command: fn(&Path, &str) -> Result<String>,
+) -> Result<BTreeMap<PackageId, Features>> {
+  let mut triples: BTreeSet<String> = BTreeSet::new();
+  if let Some(target) = settings.target.clone() {
+    triples.insert(target);
+  }
+  if let Some(targets) = settings.targets.clone() {
+    triples.extend(targets);
+  }
+
+  // Map of PackageIds using the keys that cargo-tree provides
+  let mut package_map: HashMap<(String, Version), PackageId> = HashMap::new();
+  for package in packages.iter().cloned() {
+    package_map
+      .entry((package.name, package.version))
+      .or_insert(package.id);
+  }
+
+  let mut triple_map = BTreeMap::new();
+  for triple in triples {
+    triple_map.insert(
+      triple.clone(),
+      // TODO: This part is slow, since it runs cargo-tree per-platform.
+      packages_by_platform(
+        clean_cargo_tree_output(&command(cargo_dir, triple.as_str()).map_err(|_err| {
+          Error::new(RazeError::Generic(
+            "Failed to process cargo-tree output.".into(),
+          ))
+        })?),
+        &package_map,
+      )
+      .map_err(|_err| {
+        Error::new(RazeError::Generic(
+          "Failed to segment packages by platform.".into(),
+        ))
+      })?,
+    );
+  }
+
+  Ok(
+    transpose_keys(triple_map)
+      .into_iter()
+      .map(consolidate_features)
+      .collect(),
+  )
+}
+
+fn clean_cargo_tree_output(cargo_tree_output: &str) -> Vec<String> {
+  let mut crates: Vec<String> = vec![];
```

Comment:
Can this be changed to:
```suggestion
  let mut crates = Vec::<String>::new();
```
or even just
```suggestion
  let mut crates = Vec::new();
```
?